src/fairseq2/cluster.py

```diff
@@ -182,3 +186,163 @@ def supports_current_cluster(self) -> bool:
+        self.ready_workers = 0
+        self.total_workers = num_nodes * gpus_per_node
+        self.placement_group_ids = placement_group_ids
+
+    def register_worker(self, worker_id, hostname, node_id, placement_group_id, gpu_id):
+        """Register a worker with its placement group ID and GPU ID"""
+        assert (
+            placement_group_id in self.placement_group_ids
+        ), f"Placement group id {placement_group_id} should in {self.placement_group_ids}"
+        assert 0 <= gpu_id <= 7, f"Bad gpu_id {gpu_id}"
+        node_index = self.placement_group_ids.index(placement_group_id)
+        global_rank = node_index * self.gpus_per_node + gpu_id
+        info = {
+            "worker_id": worker_id,
+            "hostname": hostname,
+            "node_id": node_id,
+            "local_rank": gpu_id,
+            "placement_group_id": placement_group_id,
+            "global_rank": global_rank,
+            "world_size": self.total_workers,
+            "local_world_size": self.gpus_per_node,
+        }
+        self.worker_info[worker_id] = info
+
+        self.ready_workers += 1
+
+        is_leader = node_index == 0 and gpu_id == 0
+
+        if is_leader:
+            self.leader_port = str(Random(self.job_id).randint(20_000, 60_000))
+            self.leader = info
+
+        return {
+            **self.worker_info[worker_id],
+            "is_leader": is_leader,
+            "ready_workers": self.ready_workers,
+            "total_workers": self.total_workers,
+        }
+
+    def get_leader_info(self):
+        if self.leader and self.ready_workers == self.total_workers:
+            result = self.leader.copy()
+            result["leader_port"] = self.leader_port
+            return result
+        else:
+            return None
+
+
+@final
+class RayClusterHandler(ClusterHandler):
+    _job_id: int | None
+    _env: MutableMapping[str, str]
+
+    def __init__(self, env: MutableMapping[str, str]) -> None:
+        self._job_id = None
+        self._env = env
+
+    @override
+    def set_torch_distributed_variables(self) -> None:
+        job_id = self._ensure_job_id()
+
+        env = self._env
+
+        ray_context = ray.get_runtime_context()
+        self.actor_id = ray_context.get_actor_id()
+        self.worker_id = str(self.actor_id)
+        self.node_id = ray_context.get_node_id()
+        self.hostname = socket.gethostname()
+        self.placement_group_id = ray_context.get_placement_group_id()
+
+        # Get the coordinator name from environment variable
+        coordinator_name = env.get(RayCoordinator.NAME)
+        assert coordinator_name
+        self.coordinator = ray.get_actor(*coordinator_name.split(":"))
+
+        accelerator_ids = ray_context.get_accelerator_ids()
+        if (
+            not accelerator_ids
+            or "GPU" not in accelerator_ids
+            or not accelerator_ids["GPU"]
+        ):
+            raise ValueError("No GPU accelerator ID available")
```

Comment:
can we just skip in that case ? to avoid ray cluster with mixed type of deployment ?

Reply:
no longer need to detect whether gpu == 0.